flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/DummyStreamExecutionEnvironment.java

```diff
@@ -0,0 +1,289 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.utils;
+
+import org.apache.flink.api.common.ExecutionConfig;
+import org.apache.flink.api.common.JobExecutionResult;
+import org.apache.flink.api.common.cache.DistributedCache;
+import org.apache.flink.api.common.restartstrategy.RestartStrategies;
+import org.apache.flink.api.dag.Transformation;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.state.AbstractStateBackend;
+import org.apache.flink.runtime.state.StateBackend;
+import org.apache.flink.streaming.api.CheckpointingMode;
+import org.apache.flink.streaming.api.TimeCharacteristic;
+import org.apache.flink.streaming.api.environment.CheckpointConfig;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.graph.StreamGraph;
+import org.apache.flink.table.planner.delegation.PlannerBase;
+
+import com.esotericsoftware.kryo.Serializer;
+
+import java.io.Serializable;
+import java.util.List;
+
+/**
+ * This is dummy {@link StreamExecutionEnvironment}, only used for {@link PlannerBase#explain(List, boolean)} method.
+ *
+ * <P>{@link Transformation}s will be added into a {@link StreamExecutionEnvironment} when translating ExecNode to plan,
+ * and they will be cleared only when calling {@link StreamExecutionEnvironment#execute()} method.
+ *
+ * <p>{@link PlannerBase#explain(List, boolean)} method will not only print logical plan but also execution plan,
+ * translating will happen in explain method. If calling explain method before execute method, the transformations in
+ * StreamExecutionEnvironment is dirty, and execution result may be incorrect.
+ *
+ * <p> All set methods (e.g. setXXX, enableXXX, disableXXX, etc) are disabled to prohibit changing configuration,
+ * all get methods (e.g. getXXX, isXXX) will be delegated to real StreamExecutionEnvironment.
+ * `execute`, `getStreamGraph`, `getExecutionPlan` methods are also disabled.
+ */
+public class DummyStreamExecutionEnvironment extends StreamExecutionEnvironment {
```

Comment:
Do I understand it correctly the only reason we need it is because of the `TableSource` interface? Because it will add transformations in the `getDataStream(StreamTransformation)` method.

If that's the case could we add a note in the javadoc that we should get rid of this class once the `TableSource` interface is reworked?

Reply:
Why not create a new instance of `StreamExecutionEnvironment`? It's not easy to maintain such a huge thing. If a new setter/getter is added to `StreamExecutionEnvironment`, it is very possible to miss overriding in this class.

Reply:
> Do I understand it correctly the only reason we need it is because of the `TableSource` interface? Because it will add transformations in the `getDataStream(StreamTransformation)` method.
> 
> If that's the case could we add a note in the javadoc that we should get rid of this class once the `TableSource` interface is reworked?

yes, it is correct. And `consumeDataStream(DataStream<T> dataStream` or `mitDataStream(DataStream<T> dataStream)` will also add transformations into execEnv.

ok, i will add the javadoc.

Reply:
> Why not create a new instance of `StreamExecutionEnvironment`? It's not easy to maintain such a huge thing. If a new setter/getter is added to `StreamExecutionEnvironment`, it is very possible to miss overriding in this class.

It's hard to create a new instance of `StreamExecutionEnvironment` with original `ExecutionConfig`. `ExecutionConfig` also has too many attributes, and is not easy to be maintained(maybe add a new attribute).

Reply:
But will we have duplicate source operators if invoking `explain` before `execute`?

Reply:
@wuchong is right. I actually thought that was the original reason for adding that class, so that we do not add transformations to the original `StreamExecutionEnvironment` (which was obviously a mistake of mine).

Reply:
Sorry, I find that the `DummyStreamExecutionEnvironment` doesn't override `addOperator(Transformation)`, so the transformations are not added to the original `env`. So I think it's correct, but easy to mis-understand ;)

Reply:
The `explain` will add the `Transformation`s to the `DummyStreamExecutionEnvironment`, that does not affect the original `StreamExecutionEnvironment`.

Reply:
So actually I was right at the very beginning :man_facepalming:

I'm fine with how it is for now. Lets maybe make sure it is well documented what are the assumptions on that class and why was it introduced.